torchrl/data/tensor_specs.py

```diff
@@ -1014,61 +1026,80 @@ class MultiDiscreteTensorSpec(DiscreteTensorSpec):
 
     def __init__(
         self,
-        nvec: Sequence[int],
+        nvec: Union[Sequence[int], torch.Tensor, int],
         device: Optional[DEVICE_TYPING] = None,
         dtype: Optional[Union[str, torch.dtype]] = torch.long,
     ):
+        if not isinstance(nvec, torch.Tensor):
+            nvec = torch.tensor(nvec)
+        if nvec.ndim < 1:
+            nvec = nvec.unsqueeze(0)
+        self.nvec = nvec
         dtype, device = _default_dtype_and_device(dtype, device)
-        self._size = len(nvec)
-        shape = torch.Size([self._size])
-        space = BoxList([DiscreteBox(n) for n in nvec])
+        shape = nvec.shape
+        space = BoxList.from_nvec(nvec)
         super(DiscreteTensorSpec, self).__init__(
             shape, space, device, dtype, domain="discrete"
         )
 
+    def _rand(self, space: Box, shape: torch.Size):
+        x = []
+        for _s in space:
+            if isinstance(_s, BoxList):
+                x.append(self._rand(_s, shape))
+            else:
+                x.append(
+                    torch.randint(
+                        0,
+                        _s.n,
+                        shape,
+                        device=self.device,
+                        dtype=self.dtype,
+                    )
+                )
+        return torch.stack(x, -1)
+
     def rand(self, shape: Optional[torch.Size] = None) -> torch.Tensor:
         if shape is None:
             shape = torch.Size([])
-        x = torch.cat(
-            [
-                torch.randint(
-                    0,
-                    space.n,
-                    torch.Size([1, *shape]),
-                    device=self.device,
-                    dtype=self.dtype,
-                )
-                for space in self.space
-            ]
-        ).squeeze()
-        _size = [self._size] if self._size > 1 else []
-        return x.T.reshape([*shape, *_size])
 
-    def _split(self, val: torch.Tensor):
-        return [val] if self._size < 2 else val.split(1, -1)
+        x = self._rand(self.space, shape)
+        if self.nvec.ndim > 1:
+            x = x.transpose(len(shape), -1)
```

Comment:
What does this do? Can you explain?

say `spec.shape = [3, 4]` and `shape=[1, 2]`
you want `x` to have shape `[1, 2, 3, 4]`. From what I understand the output of `_rand` has that shape already.
Why do you invert the dim with size `4` with the one with size `2`?

Reply:
No, in this case x have a shape of `torch.Size([1, 2, 4, 3])` without the transpose. So I have to invert 4 and 3 (no the one with size 2, because index start with 0).

This is the log of the _rand algorithm (after and before stacking):

```python
[torch.Size([1, 2]), torch.Size([1, 2]), torch.Size([1, 2]), torch.Size([1, 2])]
torch.Size([1, 2, 4])
[torch.Size([1, 2]), torch.Size([1, 2]), torch.Size([1, 2]), torch.Size([1, 2])]
torch.Size([1, 2, 4])
[torch.Size([1, 2]), torch.Size([1, 2]), torch.Size([1, 2]), torch.Size([1, 2])]
torch.Size([1, 2, 4])
[torch.Size([1, 2, 4]), torch.Size([1, 2, 4]), torch.Size([1, 2, 4])]
torch.Size([1, 2, 4, 3])
```

Because at the end I should stack in -2 dim

What do you think about this solution:

```python
    def _rand(self, space: Box, shape: torch.Size, i: int):
		....
                x.append(self._rand(_s, shape, i -1))
		....
        return torch.stack(x, -i)
```

Instead stacking in the last dimension I stack in a dimension according to the depth in the recursive box list, I pass the test with this solution.        

@vmoens